src/hotspot/share/opto/subnode.cpp

```diff
@@ -1467,27 +1467,77 @@ static bool is_counted_loop_cmp(Node *cmp) {
          n->in(0)->as_CountedLoop()->phi() == n;
 }
 
+static Node* get_reverse_cmp(int cmp_op, Node* cmp1, Node* cmp2) {
+  for (DUIterator_Fast imax, i = cmp1->fast_outs(imax); i < imax; i++) {
+    Node* u = cmp1->fast_out(i);
+    if (u->Opcode() == cmp_op && u->in(1) == cmp2 && u->in(2) == cmp1) {
+      return u;
+    }
+  }
+  return nullptr;
+}
+
+static bool is_arithmetic_cmp(Node* cmp) {
+  if (!cmp->is_Cmp()) {
+    return false;
+  }
+  const int cop = cmp->Opcode();
+  if (cop == Op_FastLock || cop == Op_FastUnlock ||
+      cop == Op_SubTypeCheck || cop == Op_VectorTest ||
+      cmp->is_Overflow()) {
+    return false;
+  }
+  return true;
+}
+
+Node* BoolNode::Identity(PhaseGVN* phase) {
+  // "Bool (CmpX a b)" is equivalent to "Bool (CmpX b a)"
+  Node *cmp = in(1);
+  if (!is_arithmetic_cmp(cmp)) {
+    return this;
+  }
+  if (phase->is_IterGVN() && outcnt() == 0) {
+    // During parsing, empty uses of bool is tolerable. During iterative GVN,
+    // we don't aggressively replace bool whose use is empty with existing node.
+    return this;
+  }
+  Node* cmp1 = cmp->in(1);
+  Node* cmp2 = cmp->in(2);
+  Node* reverse_cmp = NULL;
+  if ((_test._test == BoolTest::eq || _test._test == BoolTest::ne) &&
+      (reverse_cmp = get_reverse_cmp(cmp->Opcode(), cmp1, cmp2)) != nullptr) {
+    for (DUIterator_Fast imax, i = reverse_cmp->fast_outs(imax); i < imax; i++) {
+      Node* out = reverse_cmp->fast_out(i);
+      if (out->is_Bool() && out->as_Bool()->_test._test == _test._test &&
+          phase->type_or_null(out) != nullptr) {
```

Comment:
Why is the `phase->type_or_null` required?

Reply:
There is a cyclic case, we should avoid it.

Apply PhaseGVN for Bool A -> Generate Bool B in BoolNode::Ideal-> Apply Identity for B -> Find A while type of A is not set.